### ISSUE TITLE
[p5.js 2.0 RFC Proposal]: A visual loading indicator in place of the current "Loading" text 

### ISSUE THREAD

**Issue by SableRaf** [processing/p5.js#6795]
### Increasing access

This proposal aims to make p5.js more accessible and user-friendly, particularly for non-English speakers, by replacing the text-based "Loading" message with a simple loading animation.

### Which types of changes would be made?

- [ ] Breaking change (Add-on libraries or sketches will work differently even if their code stays the same.)
- [ ] Systemic change (Many features or contributor workflows will be affected.)
- [X] Overdue change (Modifications will be made that have been desirable for a long time.)
- [ ] Unsure (The community can help to determine the type of change.)

### Most appropriate sub-area of p5.js?

- [ ] Accessibility
- [ ] Color
- [X] Core/Environment/Rendering
- [ ] Data
- [X] DOM
- [ ] Events
- [ ] Image
- [ ] IO
- [ ] Math
- [ ] Typography
- [ ] Utilities
- [ ] WebGL
- [ ] Build process
- [ ] Unit testing
- [X] Internationalization
- [ ] Friendly errors
- [ ] Other (specify if possible)

### What's the problem?

Currently, p5.js displays a text-based "Loading" message in the top-left corner of the page during the execution of the `preload()` function. This approach is not very visually engaging and can be confusing for end-users who do not understand English or have visual impairments. 

A p5.js coder can replace "Loading" with a custom animation. This is mentioned in the [`preload()`](https://p5js.org/reference/#/p5/preload) and the [p5.js overview page](https://github.com/processing/p5.js/wiki/p5.js-overview#loading-screen) but this option is still relatively hidden and not particularly beginner friendly.  See this [Coding Train tutorial on loading animations](https://www.youtube.com/watch?v=UWgDKtvnjIU).

### What's the solution?

Replace the text-based "Loading" message with a visual loading animation. This animation could be a simple, universally recognized loading indicator (e.g., spinning circle) that provides a visual cue to users that content is being loaded without relying on text. An animated loader also better conveys the fact that _something is happening_. We can use the [`afterPreload` lifecycle event](https://github.com/processing/p5.js/issues/6422) to trigger hiding the loading animation.

The loading animation could itself be made with p5.js and editable by the user.

### Pros (updated based on community comments)

**Accessibility:** Enhances p5.js accessibility by providing a language-independent loading indicator.
**User Experience:** Improves the user experience with a more engaging and informative visual cue during load times.
**Internationalization:** Supports p5.js's goal of being globally accessible by removing language barriers in the loading process.
**Visual identity:** The default loading animation could be an opportunity to include subtle elements of the p5.js visual identity like the pink color.

### Cons (updated based on community comments)

**Intrusivity:** Depending on its design, the loading animation may be perceived as intrusive. We can mitigate this by making it easy to change the animation or deactivate it, and add examples on the `preload()` reference page.

### Proposal status

Under review

**Comment by perminder-17** [comment]
Thank you for bringing this up @SableRaf . I find your proposal satisfactory, and all the mentioned benefits make sense. Initially, I thought using text for loading, instead of animation, was to prevent confusion for users creating sketches and animations in p5.js. 

I apologize if I am wrong but do you think using a loading animation might give a similar appearance to a sketch, causing potential confusion.

**Comment by haroon10725** [comment]
@davepagurek Can I work on this issue?

**Comment by limzykenneth** [comment]
@SableRaf If we adopt async/await per #6767 would a loading indicator still be needed? With async/await there will not be a need for a separate `preload()` function and data loading probably will feel similar to the synchronous behaviour of Processing for example.

**Comment by SableRaf** [comment]
I'm not sure I fully understand the implications of using async/await for loading so take my feedback with a pinch of salt but, in cases where large files or many assets are being loaded, would a loading indicator could still be beneficial to inform users that the process is ongoing? Maybe it can be less intrusive, like a loading spinner in the top left corner.

**Comment by limzykenneth** [comment]
@SableRaf I was thinking something along those lines if we are to still retain the loading indicator with async/await. I believe that is possible, just wondering if it is something we want to keep around. I don't quite remember but does Processing do anything similar when waiting for slow file reads?

**Comment by SableRaf** [comment]
> does Processing do anything similar when waiting for slow file reads?

Not that I'm aware, but file reads are probably faster with Java on Desktop, right?

**Comment by limzykenneth** [comment]
Yes, certainly with file read from disk definitely faster than file request over the network. So I think there are a few options:

* Full canvas blocking loading indicator - Similar to what we have now, perhaps visual instead, cons would be that it can seem intrusive
* Loading spinner/graphic in corner - Less intrusive (although not entirely unintrusive)
* Opt-in or Opt out for the above two - As part of the loading API, sketch author can decide whether to use the loading indicator or not through an additional argument. It provides choice but slightly complicates the API, also it is unclear whether loading indicator should be opt in or opt out
* No loading indicator at all - Likely closest to semantic JavaScript but long loading time can be misunderstood as performance issue

**Comment by SableRaf** [comment]
Thanks @limzykenneth, that sums it up perfectly!

I just thought of a way to make this work as opt-in in a relatively non-intrusive way: what if when files take a long time to load, we throw a warning message via the FES suggesting that the sketch author use the optional loader? For example:

> 🌸 It looks like your file is taking a bit longer to load than usual ⏳ Consider using `loadingAnimation()` to display a loading animation while your files are being loaded.

This way people would only be exposed to the feature when they need it.

**Comment by quinton-ashley** [comment]
In response to the opt-in idea from @SableRaf, if developers are loading assets locally or from the browser's cache, they wouldn't experience long loading times (unless they throttle with dev tools), so they wouldn't see the message suggesting to enable p5's loading animation.

> No loading indicator at all - Likely closest to semantic JavaScript but long loading time can be misunderstood as performance issue

Is this the current behavior of p5 v2? That's fine in a majority of user experiences where a loading indicator is not needed, but not everyone has fast internet. If the responsibility falls to p5 dev users to provide their own loading screen, they may choose not to make one based on their own perception of how fast the sketch "should" load.

@perminder-17 brings up an interesting point

> do you think using a loading animation might give a similar appearance to a sketch, causing potential confusion??

Yes, so I'd be nice to have an opt-out CSS loading animation, something simple like the p5 asterisk spinning above the canvas, still with "Loading..." text included, that would run until async setup is finished. Also the cursor could be changed to indicate loading too.

Although with async setup, users could easily display something static during loading, it'd also be good to offer updated tutorials on p5js.org that detail how users could opt-out of the default loading animation and implement their own custom loading animations. The tutorial could be suggested by p5 FES to users that use a lot of load* functions in their sketch. Something like this:

```js
let finishedLoading = false;
let dog, cat;

async function setup() {
  createCanvas(200, 200);

  customLoader();
  // setup finishes immediately and
  // the default loading animation is not shown
}

async function customLoader() {
  dog = await loadImage('dog.png');
  cat = await loadImage('cat.png');

  // simulate long loading

  finishedLoading = true;
}

function draw() {
  if (!finishedLoading) {
   // show custom loading animation
   return;
  }
  // rest of the sketch that uses loaded assets
}
```

**Comment by ksen0** [comment]
Thanks for the discussion and ideas everyone, it would be great to see this implemented. **This proposal has been accepted for implementation and is looking for a volunteer!** This is a nice stand-alone feature, so it would be great to keep this open for new contributors - if you're already an established contributor, please check first if there are other issues ["Ready for Implementation" in p5.js 2.x](https://github.com/orgs/processing/projects/21/views/2).

Solutions can include either/or:

1. Tutorial for beta.p5js.org (on p5.js-website 2.0 branch) that shows how to create a custom loading animation with async. For example, a background color and a simple text, which the user can control. As @quinton-ashley suggests, "The tutorial could be suggested by p5 FES to users that use a lot of load* functions in their sketch."
2. A CSS loading animation with cursor effects **add-on library,** which can provides `loadingAnimation` as a way to parameterize CSS loading. This would be a very small library, but by making it separately first, it makes it possible to test it, iterate on it, and discuss it before making core API changes. 

**If you're new to p5.js contribution, welcome!** Please be sure to read through the [contributor guidelines](https://p5js.org/contribute/contributor_guidelines/), and keep in mind that you should not file a pull request (or start working on code changes) without a corresponding issue or before an issue has been approved for implementation and assigned.

_PS: This message is part of cleaning up the [p5.js 2.0 roadmap](https://github.com/orgs/processing/projects/21) prior to release, to make outstanding more more approachable for new contributors. You can join the conversation about this new version on [this Discourse thread](https://discourse.processing.org/t/dev-updates-p5-js-2-0-you-are-here/46130) or this [GitHub thread](https://github.com/processing/p5.js/issues/7488)!_

**Comment by SableRaf** [comment]
Hi @ksen0 and thanks for the summary!

On the tutorial idea, I've updated a 1.x demo sketch of mine for async/await [here](https://editor.p5js.org/SableRaf/sketches/Nb0WPmyI6). It could potentially be used as a basis for a tutorial on custom loaders. I'm curious if that's the direction you had in mind.

About the loading text idea, I'm wondering if it’s necessary? Visual loading indicators (not necessarily a spinning star) are pretty common, and I think most people would recognize them without text. But I know you had some thoughts about this, so I’m open to whatever makes most sense 😌

**Comment by quinton-ashley** [comment]
>  For example, rather than a spinning star animation, it could be a very simple animation of some "loading" text in many languages? Maybe it can be more creative than "loading?" 🌐

@ksen0 I like your idea! Including a spinning p5.js asterisk, the p5.js logo, or some other kind of branding would help to make it clear what's loading though.

The mouse cursor could also be set to indicate loading when hovering over the canvas that's loading.

@SableRaf The concern with a text-less loader from @perminder-17 was that people may confuse it for the art itself in cases where that could be initially ambiguous. Would you disagree?

Also I was under the impression that the loader would just cover individual p5 canvases and not the full page.

If we assume p5 users want viewers to be able to see their canvas content as soon as possible maybe just cutting to it or having a very short fade out may be preferable.

> it would be great to keep this open for new contributors

I have some other suggestions for anyone that wants to implement this feature:

- we don't need users to see the loading animation at all if the sketch loads fast enough
- ideally a CSS loader has to be ready to go immediately, if it uses any images they need to be base64 embedded in the p5.js file

**Comment by perminder-17** [comment]
> the mouse cursor could also be set to indicate loading when hovering over the canvas that's loading

I loved this idea, this could be great and we should consider a mouse cursor to be also set to indicate loading while hovering the canvas.

**Comment by quinton-ashley** [comment]
Could be cool to have the mouse cursor become a spinning p5.js asterisk!

Partly I think this issue is trying to solve the problem of having a totally blank canvas during loading though.

**Comment by AhmedMagedC** [comment]
Hi, so I'm a new contributor I'm not yet familiar with the codebase
how can i get this issue assigned to me? do i need to provide basic ideas in advance?

**Comment by ksen0** [comment]
HI @AhmedMagedC welcome! There's a few different ideas here, hav you had a chance to check [the earlier comment](https://github.com/processing/p5.js/issues/6795#issuecomment-2809160037) out? From that and all the later comments and ideas, what do you think is most interesting and doable to you? To recap: the 2 options are to write a tutorial or to create an add-on library. The add-on library can explore any of the ideas listed here (cursor override, various loading text, etc), it's an open topic and when the first version is done others can iterate on it.

**Comment by AhmedMagedC** [comment]
I think the first one is more doable to me
If i understand the idea correctly, when setup function being async and when the loading takes too long (Or maybe many assets are being loaded ?)
a tutorial is suggested by p5 FES to users showing them how to create their own custom loading animation, so
my responsibility is to write that tutorial and to show that message when the loading takes too long?

**Comment by SableRaf** [comment]
Hi @ksen0 and all,

I took the liberty of putting together a proof of concept based on Kit's suggestions. It includes a randomized list of "Loading..." words in multiple languages, and an (optional) spinning p5.js logo for the cursor. Let me know if that's close to what you had in mind.

The code is available on OpenProcessing here: https://openprocessing.org/sketch/2619980

Thanks to @Stefterv for helping me figure out the lifecycle hooks in p5.js 2.0, and how to use p5.js for the loading animation (instead of CSS).

![Image](https://github.com/user-attachments/assets/446c7ca8-f759-4a57-90a5-47191b12da60)

**Comment by VANSH3104** [comment]
@SableRaf, your loading animation is fantastic! The p5-branded spinner and multilingual support make this an elegant solution. Turning it into an add-on could be a brilliant idea

**Comment by AhmedMagedC** [comment]
> Hi [@ksen0](https://github.com/ksen0) and thanks for the summary!
> 
> On the tutorial idea, I've updated a 1.x demo sketch of mine for async/await [here](https://editor.p5js.org/SableRaf/sketches/Nb0WPmyI6). It could potentially be used as a basis for a tutorial on custom loaders. I'm curious if that's the direction you had in mind.
> 
> About the loading text idea, I'm wondering if it’s necessary? Visual loading indicators (not necessarily a spinning star) are pretty common, and I think most people would recognize them without text. But I know you had some thoughts about this, so I’m open to whatever makes most sense 😌

Hi [SableRaf](https://github.com/SableRaf),
Thanks for the demo sketch! it could indeed be a good basis to helping writing the tutorial.
but I had an issue running your demo in 1.x versions, the image doesn't display after loading animation is done, when i tried to debug i found that draw function starts immediately, it doesn't wait for the async functions in setup to be awaited
so is this a normal behavior? also giving that it works normally in 2.0.0 version (draw function doesn't start until all async functions in setup is completed)

**Comment by AhmedMagedC** [comment]
Hi [@ksen0 ](https://github.com/ksen0), how are you?
Just checking in to inform you I'm still interested in writing the tutorial, honestly I'm also interested in implementing the add-on library, but after finishing the tutorial

# How to approach it?
so after reading through the comments, I think there are two main approaches for the user to create custom loading animation
1. using p5 visual elements to create it like what was mentioned by [@quinton-ashley ](https://github.com/quinton-ashley), also was done by this tutorial [9.14: Loading Animation - p5.js Tutorial](https://www.youtube.com/watch?v=UWgDKtvnjIU)

2. using CSS to create the animation, taking this [demo sketch](https://editor.p5js.org/SableRaf/sketches/Nb0WPmyI6) by [@SableRaf ](https://github.com/SableRaf) as a strong reference

so should i create new issue on p5.js-website?

also any feedback would be highly appreciated ❤️

**Comment by SableRaf** [comment]
Hi @AhmedMagedC. I would recommend building on the approach I shared in my latest reply (and [this proof of concept](https://openprocessing.org/sketch/2619980)) rather than the earlier CSS-based prototype. This version is more aligned with p5.js v2.0 (using async/await) and relies fully on p5.js for rendering. It also has the advantage that the loader starts after the p5 canvas is created, so it can match the correct size. I understand the code might feel a bit more complex, but I'm happy to walk through it together if that would be helpful! 

PS: @ksen0 let me know if that approach makes sense to you as well or if you have a different idea

**Comment by AhmedMagedC** [comment]
[@ksen0](https://github.com/ksen0) Thanks for providing the bigger picture, I now fully understand what to do.

[@SableRaf](https://github.com/SableRaf) Thanks for providing this example, i will sure contact with you if i need help.

as for now i will preferably work on the add-on library 

so [@ksen0](https://github.com/ksen0) can you assign this issue to me?

**Comment by ksen0** [comment]
Thanks @AhmedMagedC , add-on library sounds great, assigned!

**Comment by AhmedMagedC** [comment]
Hi guys, how are you?
So I would like to keep you updated !

as for now I have created the addon library and it works as expected with p5 2.x version

but i have issues making the lib compatible with 1.x , while assets are loading in `preload()`canvas and p5 visual elements are not yet initialized, making it impossible to depend on p5 elements to create the loading animation/spinning logo

so what do you think i should do? 

also another question when it comes to coding best practice, can i use internal p5 variables that was intended for internal use only like `_setupDone` or should i find workaround instead?

[repo](https://github.com/AhmedMagedC/p5.loadingAnimation.js)

**Comment by VANSH3104** [comment]
Hi @AhmedMagedC , Great work, it looks good! 👏. Could you also share a short demo video like @SableRaf  did? It would help us see it in action. About p5 1.x support: I think you're right that the preload() phase can be tricky since the canvas and visual elements aren't fully initialized yet. Since preload() runs before canvas init, maybe consider a standalone loader outside p5. Also, using internal vars like _setupDone isn’t ideal for stability; it's better to use custom flags, if possible.

**Comment by AhmedMagedC** [comment]
> Hi [@AhmedMagedC](https://github.com/AhmedMagedC) , Great work, it looks good! . Could you also share a short demo video like [@SableRaf](https://github.com/SableRaf) did? It would help us see it in action. About p5 1.x support: I think you're right that the preload() phase can be tricky since the canvas and visual elements aren't fully initialized yet. Since preload() runs before canvas init, maybe consider a standalone loader outside p5. Also, using internal vars like _setupDone isn’t ideal for stability; it's better to use custom flags, if possible I think.

Hi [@VANSH3104](https://github.com/VANSH3104), thanks for the suggestion
I will surely add examples and docs in my repo when i finish the logic itself

Can you please elaborate more on the idea of standalone loader outside p5?
do you mean to make it depends on CSS ?

**Comment by VANSH3104** [comment]
Hi @AhmedMagedC , Just to clarify by "standalone loader outside p5", I didn’t mean a CSS-based one. Instead, I was referring to a loader built with p5.js itself, but initialized after the canvas is created, outside of the preload() lifecycle. That way, we avoid the limitations of preload().

**Comment by AhmedMagedC** [comment]
> Hi [@AhmedMagedC](https://github.com/AhmedMagedC) , Just to clarify, by "standalone loader outside p5", I didn’t mean a CSS-based one. Instead, I was referring to a loader built with p5.js itself, but initialized after the canvas is created, outside of the preload() lifecycle. That way, we avoid the limitations of preload(). You can see SableRaf solution.

my solution is already built on SableRaf's solution, which works on 2.x 
 
but with 1.x p5 version, assets needs to be loading in `preload()` in which p5 visual elements are not yet ready and cannot be used to create the loading animation/spinning logo

[@ksen0](https://github.com/ksen0) [@SableRaf](https://github.com/SableRaf), kinda stuck here if you can help! I think i can add CSS-based solution when user works with `preload()` in 1.x p5 version , so what do you think?

**Comment by limzykenneth** [comment]
@AhmedMagedC We are only looking to implement this in 2.x. 1.x has completely different requirements and we are currently not accepting any new feature to 1.x, instead focusing on 2.x more, so if there is a solution for 2.x already it is good enough for this issue.

**Comment by Anirudh-x** [comment]
Hi! I've submitted a PR to address this issue: #8261 

Summary:
- Replaced text "Loading..." with an animated CSS spinner
- Used p5.js brand color (#ED225D)
- Smooth rotation animation with flexbox centering
- Maintained customization options via CSS

The spinner displays during asset loading in `preload()` and removes automatically when loading completes. I've also created a comprehensive manual test suite to verify functionality across different scenarios.

Would appreciate any feedback! 🙂

**Comment by harshiltewari2004** [comment]
Hi i would like to work on this issue

**Comment by 10billionpercent** [comment]
Hi! I was exploring the p5.js editor in dark mode and noticed a readability / contrast issue with the loading indicator.
The “Loading…” text in the Preview panel appears very low-contrast (dark text on dark background), which makes it hard to notice whether the sketch is loading or stuck.
This happens consistently in dark mode and is reproducible both on desktop and mobile (including desktop view on mobile).
Attaching screenshots for context — one full view and one cropped close-up highlighting the issue.
This seems like a small accessibility improvement independent of changing the loading indicator to an animation. Happy to help further if this is useful.
![image (3).png](https://github.com/user-attachments/assets/b965ec95-4f35-429f-82eb-af28156305e1)
![image (2).png](https://github.com/user-attachments/assets/99ebc25b-5ea6-4c38-930a-ceeba251eed1)

**Comment by ksen0** [comment]
Sorry this has been stuck in review for such a while. The reason was that the add-on that was submitted had introduced unexpected API usage and I got stuck unsure how to provide productive feedback, so I am sorry for this. This can be picked up again; but separately, I've created an issue https://github.com/processing/p5.js/issues/8922 for the specific issue of loading indication being missing in p5.js 2.x as that is a separate issue than the improvements for the indicator discussed here. Apologies and thanks for understanding. I hope that at least separating these 2 topics can be helpful.

**Comment by rizzhal** [comment]
is someone still working on this issue ? If not I want to work on it

**Comment by ksen0** [comment]
The codeday.org team of @Jextic and co said last week they were working on a follow up for this topic. @Jextic please share a status update here in the next week? Doesn't need to be done, but as an update. If there's no response by end of day August 21, the issue may be opened to volunteers/reassigned
